### ISSUE TITLE
`Appsignal::Logger` breaks Rails' stdout logger broadcast feature

### ISSUE THREAD

**Issue by tbuehlmann** [appsignal/appsignal-ruby#930]
Learned about Appsignal's new Logging feature today and fiddled with it, exciting! Had it online on a staging environment real quick and it was working, but the rails server wouldn't log to stdout anymore, which was the case before switching things.

Turns out, if you run a rails console or a rails server started using `bin/rails server` (using `--log_to_stdout` or in development), rails will add a broadcast to `Rails.logger` that will let it log to stdout as well as what `Rails.logger` logs to.

Rails basically does this:

```ruby
console = ActiveSupport::Logger.new(STDOUT)
console.formatter = Rails.logger.formatter
console.level = Rails.logger.level

unless ActiveSupport::Logger.logger_outputs_to?(Rails.logger, STDERR, STDOUT)
  Rails.logger.extend(ActiveSupport::Logger.broadcast(console))
end
```

… where `ActiveSupport::Logger.broadcast` returns a module that defines `#add`:

```ruby
Module.new do
  define_method(:add) do |*args, &block|
    logger.add(*args, &block)
    super(*args, &block)
  end

  …
end
```

So `Rails.logger.add(…)` will log to stdout and then call the logger's regular `add` method. **It will only do this if the `#add` method is used**. Now the bug:

It doesn't work with `Appsignal::Logger` as `#info`, `#debug` and the other severity methods **don't internally call `#add`**. Example:

```ruby
def info(message = nil, attributes = {})
  return if INFO < level
  message = yield if message.nil? && block_given?
  return if message.nil?
  Appsignal::Extension.log(
    @group,
    3,
    message,
    Appsignal::Utils::Data.generate(attributes)
  )
end
```

It would be nice if the severity methods would call `#add` internally, too, so the broadcast is working properly. As I don't need any Appsignal attributes in my logs yet, I came up with a custom logger that works for me:

```ruby
# lib/simple_appsignal_logger.rb

class SimpleAppsignalLogger < Logger
  # From https://github.com/appsignal/appsignal-ruby/blob/9f77f6446e2c93ab2b073608fff1135f7229d757/lib/appsignal/logger.rb#L36-L49
  APPSIGNAL_SEVERITIES = Hash.new(0).tap do |severities|
    severities[DEBUG] = 2
    severities[INFO] = 3
    severities[WARN] = 5
    severities[ERROR] = 6
    severities[FATAL] = 7
  end

  APPSIGNAL_ATTRIBUTES = Appsignal::Utils::Data.generate({})

  def initialize(group, level: DEBUG)
    @group = group
    self.level = level
  end

  def add(severity, message = nil, progname = nil)
    severity ||= UNKNOWN

    if severity < level
      return true
    end

    if message.nil?
      if block_given?
        message = yield
      else
        message = progname
      end
    end

    if message
      Appsignal::Extension.log(@group, APPSIGNAL_SEVERITIES[severity], message, APPSIGNAL_ATTRIBUTES)
      true
    else
      nil
    end
  end
end
```

```ruby
# config/environments/staging.rb

require_relative "../../lib/simple_appsignal_logger"

Rails.application.configure do
  # …

  config.log_formatter = ::Logger::Formatter.new
  logger = SimpleAppsignalLogger.new("rails")
  logger.formatter = config.log_formatter
  config.logger    = ActiveSupport::TaggedLogging.new(logger)
end
```

**Comment by shairyar** [comment]
https://app.intercom.com/a/inbox/yzor8gyw/inbox/admin/4356044/conversation/16410700202250 (private link)

**Comment by luismiramirez** [comment]
Thanks for the detailed report, @tbuehlmann! You can follow the fixing progress on the linked PR

**Comment by luismiramirez** [comment]
You can find the fix on version 3.3.10 of the AppSignal gem.

**Comment by tbuehlmann** [comment]
Great, thank you!

**Comment by tbuehlmann** [comment]
Have to chime in again: We're not yet there. The method signatures differ too much. Example:

```ruby
logger = Appsignal::Logger.new("rails")
console = Logger.new(STDOUT)

logger.extend(ActiveSupport::Logger.broadcast(console))

logger.info("foo") # ArgumentError: wrong number of arguments (given 4, expected 1..3)
```

```
# Appsignal::Logger
def add(severity, message = nil, group = nil, attributes = {})

# Logger
def add(severity, message = nil, progname = nil)
```

The broadcast overrides `#add` and calls:

```ruby
define_method(:add) do |*args, &block|
  # for this example, args would be:
  # [1, "foo", "rails", {}]

  logger.add(*args, &block)
  super(*args, &block)
end
```

So that calls `Logger#add` with 4 arguments where it only accepts 1 to 3. Not sure how we could circumvent this.

**Comment by tbuehlmann** [comment]
For the moment, a customized logger does the job:

```ruby
class AppsignalCompatibleLogger < Logger
  include ActiveSupport::LoggerSilence

  # original signature:
  # add(severity, message = nil, progname = nil)
  def add(severity, message = nil, progname = nil, *_args)
    super(severity, message, progname)
  end
end
```

```ruby
logger = Appsignal::Logger.new("rails")
console = AppsignalCompatibleLogger.new(STDOUT)

logger.extend(ActiveSupport::Logger.broadcast(console))

logger.info("foo") # works as expected
```

Doesn't fix the issue ofc. Not sure how to actually fix this without introducing some kind of per-thread (or fiber) variable or changing rails code. This could work, but meh:

```ruby
class Appsignal::Logger
  def info(message = nil, attributes = {})
    return if INFO < level
    message = yield if message.nil? && block_given?
    return if message.nil?

    with_attributes(attributes) do
      add(INFO, message, @group)
    end
  end
  
  def add(severity, message = nil, group = nil)
    severity ||= UNKNOWN
    return true if severity < level
    group = @group if group.nil?
    if message.nil?
      if block_given?
        message = yield
      else
        message = progname
        group = @group
      end
    end
    return if message.nil?
    message = formatter.call(severity, Time.now, group, message) if formatter

    Appsignal::Extension.log(
      group,
      SEVERITY_MAP.fetch(severity, 0),
      @format,
      message,
      Appsignal::Utils::Data.generate(attributes)
    )
  end

  private

  def with_attributes(attributes)
    begin
      Thread.current[:appsignal_logger_attributes] = attributes
      yield
    ensure
      Thread.current[:appsignal_logger_attributes] = nil
    end
  end

  def attributes
    Thread.current.fetch(:appsignal_logger_attributes, {})
  end
end
```

Seems rather sophisticated for such a small issue, maybe you have a better idea.

**Comment by luismiramirez** [comment]
Hey @tbuehlmann! Thank you so much for your insight. I've pushed a branch that uses an instance variable instead threads. Here's the branch: https://github.com/appsignal/appsignal-ruby/pull/938

My tests went OK but I would like to know if the fix is suitable for your use case.

**Comment by tbuehlmann** [comment]
Yep, that works. But it seems to be prone to race conditions when multiple threads use the same logger, isn't it?

**Comment by luismiramirez** [comment]
> But it seems to be prone to race conditions when multiple threads use the same logger, isn't it?

Indeed. I'm currently trying an approach with mutex synchronization similar to what Ruby std `Logger` uses, but it's complicating things too much for this use case. Using thread locals as you proposed seems like the simplest solution for thread safety right now.

**Comment by luismiramirez** [comment]
Latest update on the linked branch contains the change.

**Comment by unflxw** [comment]
For reference, the final fix (in #938) was fixed in version `3.4.0`.
